### ISSUE TITLE
Transformer Encoder Layer with src_key_padding makes NaN

### ISSUE THREAD

**Issue by JunhyunB** [pytorch/pytorch#24816]
I'm using PyTorch 1.3 and this is sample code for my example.

```python
import torch
import torch.nn as nn

enc = nn.TransformerEncoderLayer(3,1)
model = nn.TransformerEncoder(enc, 6)

x = torch.Tensor([[[1,2,3],[0,5,6]],[[-1,-2,0],[-4,-5,0]],[[1,2,3],[0,5,6]],[[-1,-2,0],[-4,-5,0]]])
y = torch.tensor([[0,0],[1,0],[1,1],[1,1]]).bool()
print(x.size())
print(y.size())
print(y)
output = model(src=x.transpose(0,1), src_key_padding_mask=y).transpose(0,1)
print(output)

>> torch.Size([4, 2, 3])
torch.Size([4, 2])
tensor([[False, False],
        [ True, False],
        [ True,  True],
        [ True,  True]])
tensor([[[-0.0933, -1.1754,  1.2687],
         [-0.4194, -0.9599,  1.3794]],

        [[ 0.9560, -1.3805,  0.4246],
         [ 0.9844, -1.3715,  0.3871]],

        [[    nan,     nan,     nan],
         [    nan,     nan,     nan]],

        [[    nan,     nan,     nan],
         [    nan,     nan,     nan]]], grad_fn=<TransposeBackward0>)
```

If I have all padded sequence with padding mask, this makes NaN output.
and backward() of this makes every NaN of outputs for next step of iterations.

I tried 
```python
output[torch.isnan(output)] = 0 
```
but it wasn't helpful, every output became zero. (mean still NaN for every output value)

How can I solve this?

cc @SsnL

**Comment by ZhuBaohe** [comment]
you can try as follow:

`output = output.masked_fill(torch.isnan(output), 0)`
`print(output)`

```
>>>tensor([[[-0.0933, -1.1754,  1.2687],
            [-0.4194, -0.9599,  1.3794]],

           [[ 0.9560, -1.3805,  0.4246],
            [ 0.9844, -1.3715,  0.3871]],

           [[ 0.0000,  0.0000,  0.0000],
            [ 0.0000,  0.0000,  0.0000]],

           [[ 0.0000,  0.0000,  0.0000],
            [ 0.0000,  0.0000,  0.0000]]], grad_fn=<MaskedFillBackward0>)
```

**Comment by JunhyunB** [comment]
Your suggestion can make nan to 0 for output,
but after weight update, every output becomes nan, too..

```python
import torch
import torch.nn as nn

enc = nn.TransformerEncoderLayer(3,1)
model = nn.TransformerEncoder(enc, 6)

x = torch.Tensor([[[1,2,3],[0,5,6]],[[-1,-2,0],[-4,-5,0]],[[1,2,3],[0,5,6]],[[-1,-2,0],[-4,-5,0]]])
y = torch.tensor([[0,0],[1,0],[1,1],[1,1]]).bool()
print(x.size())
print(y.size())
print(y)
output = model(src=x.transpose(0,1), src_key_padding_mask=y).transpose(0,1)
print(output)
output = output.masked_fill(torch.isnan(output), 0)
print(output)
optim = torch.optim.Adam(model.parameters())
optim.zero_grad()
output.mean().backward()
optim.step()
output = model(src=x.transpose(0,1), src_key_padding_mask=y).transpose(0,1)
output = output.masked_fill(torch.isnan(output), 0)
print(output)
>>> torch.Size([4, 2, 3])
torch.Size([4, 2])
tensor([[False, False],
        [ True, False],
        [ True,  True],
        [ True,  True]])
tensor([[[-0.2467, -1.0826,  1.3293],
         [-0.9920, -0.3769,  1.3689]],

        [[-0.6376, -0.7744,  1.4120],
         [-0.5266, -0.8733,  1.4000]],

        [[    nan,     nan,     nan],
         [    nan,     nan,     nan]],

        [[    nan,     nan,     nan],
         [    nan,     nan,     nan]]], grad_fn=<TransposeBackward0>)
tensor([[[-0.2467, -1.0826,  1.3293],
         [-0.9920, -0.3769,  1.3689]],

        [[-0.6376, -0.7744,  1.4120],
         [-0.5266, -0.8733,  1.4000]],

        [[ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000]],

        [[ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000]]], grad_fn=<MaskedFillBackward0>)
tensor([[[0., 0., 0.],
         [0., 0., 0.]],

        [[0., 0., 0.],
         [0., 0., 0.]],

        [[0., 0., 0.],
         [0., 0., 0.]],

        [[0., 0., 0.],
         [0., 0., 0.]]], grad_fn=<MaskedFillBackward0>)

```

**Comment by ZhuBaohe** [comment]
I have commit a PR for this problem.

**Comment by zhangguanheng66** [comment]
This is not an issue related to `nn.Transformer` or `nn.MultiheadAttention`.

After the key_padding_mask filter layer, `attn_output_weights` is passed to `softmax` and here is the problem. In your case, you are fully padding the last two batches (see `y`). This results in two vectors fully filled with `-inf` in `attn_output_weights`. If a tensor fully filled with `-inf` is passed to softmax, softmax will return `nan` as a result. Will check with the team and see if this behavior is expected.

IMO, it's not correct to reset `nan` to zero in `attn_output_weights`.

A quick solution to your case. If you want to fully pad some batches, you could simply remove them, right?

**Comment by JunhyunB** [comment]
Yes, that's exactly right.
I just wonder there are another solutions rather than just erase my input.

**Comment by zhangguanheng66** [comment]
As @soumith said, it would add some perf overhead for a full pass to check the `nan` values. https://github.com/pytorch/pytorch/issues/25110#issuecomment-524519087

**Comment by zhangguanheng66** [comment]
Feel free to re-open the issue if you still have questions.

**Comment by zml24** [comment]
Still a bug in torch/nn/functional.py
<img width="705" alt="image" src="https://user-images.githubusercontent.com/17996245/67767427-87f6b000-fa8b-11e9-8dd9-bd193566978b.png">
Let's see what will happen with inf input in softmax.
<img width="563" alt="image" src="https://user-images.githubusercontent.com/17996245/67767828-3d296800-fa8c-11e9-8578-c0ecab3ab05a.png">
Can we just softmax first and mask with 0 and renormalization?
I think it's correct and effective.

**Comment by zml24** [comment]
or just
`attn_output_weights = attn_output_weights.masked_fill(key_padding_mask.unsqueeze(2).unsqueeze(3), sys.float_info.min)`

**Comment by zml24** [comment]
or just
```
attn_output_weights = attn_output_weights.masked_fill(
    key_padding_mask.unsqueeze(1).unsqueeze(2),
    sys.float_info.min)
```

**Comment by zml24** [comment]
Same bug here
<img width="765" alt="image" src="https://user-images.githubusercontent.com/17996245/67770873-3140a480-fa92-11e9-94c4-308f1b2b50a2.png">
I fixed these two bugs myself and `nn.Transformer` can finally match my transformer module's learning ability.

**Comment by zhangguanheng66** [comment]
It's not a bug in `nn.Transformer`. You should check out the master branch, and there was a type promotion issue fixed by https://github.com/pytorch/pytorch/pull/28253

**Comment by isaacmg** [comment]
I'm having the same issue but I'm not even using a softmax function (as I'm working with time series data). The following is returning NaN values after the first gradient step.
```
class MultiAttnHeadSimple(torch.nn.Module):
    def __init__(self, n_time_series, d_model=128):
        super().__init__()
        self.dense_shape = torch.nn.Linear(n_time_series, d_model)
        self.pe = SimplePositionalEncoding(d_model)
        self.multi_attn = MultiheadAttention(embed_dim=d_model, num_heads=8, dropout=0.1)
        self.final_layer = torch.nn.Linear(d_model, 1)
    def forward(self, x, mask=None):
        x = self.dense_shape(x)
        x = self.pe(x)
        x = x.permute(1,0,2)
        if mask is None:
            x = self.multi_attn(x, x, x)[0]
            print(torch.isnan(x))
        else: 
            x = self.multi_attn(x, x, x, attn_mask=mask)[0]
        
        x = self.final_layer(x)
        return x 
```

**Comment by zhangguanheng66** [comment]
Could you provide the input `x` and `mask` so we could reproduce the error?

**Comment by isaacmg** [comment]
Thanks for the response. I was able to fix it on my own. I didn't normalize my time series values before passing them in which caused the loss to explode. After applying StandardScaler prior to passing in the time series value the problem went away.

**Comment by zhangguanheng66** [comment]
> I'm having the same issue but I'm not even using a softmax function (as I'm working with time series data). The following is returning NaN values after the first gradient step.
> 
> ```python
> class MultiHeadAttnSimple(torch.nn.Module):
>     def __init__(self, n_time_series, d_model=128):
>         super().__init__()
>         self.dense_shape = torch.nn.Linear(n_time_series, d_model)
>         self.pe = SimplePositionalEncoding(d_model)
>         self.multi_attn = MultiheadAttention(embed_dim=d_model, num_heads=8, dropout=0.1)
>         self.final_layer = torch.nn.Linear(d_model, 1)
>     def forward(self, x, mask=None):
>         x = self.dense_shape(x)
>         x = self.pe(x)
>         x = x.permute(1,0,2)
>         if mask is None:
>             x = self.multi_attn(x, x, x)[0]
>             print(torch.isnan(x))
>         else: 
>             x = self.multi_attn(x, x, x, attn_mask=mask)[0]
>         
>         x = self.final_layer(x)
>         return x 
> ```

Just to clarify an early comment. There is a `softmax` function within `MultiheadAttention` function. See [here](https://github.com/pytorch/pytorch/blob/94016b153a4130814d95181014f635b33eb117df/torch/nn/functional.py#L3375)

**Comment by lrs1353281004** [comment]
I have a similar  problem while using transformer with src_key_padding_mask.(pytorch version:1.5.0, gpu)
When I use a mask tensor with type : torch.float , it works normally.
example code:
```python
import torch.nn as nn
import torch
en_layer = nn.TransformerEncoderLayer(d_model=300, nhead=2)
transformer_en = nn.TransformerEncoder(en_layer, num_layers=1)

inputs=torch.rand((10,5,300))
masks =torch.ones((5,10))

output=transformer_en(inputs,src_key_padding_mask=masks)
```
But when I use a mask tensor with type: torh.uint8(As required in the documentation of pytorch transformer) or torch.bool ,  it return output with  all nan values.
example code:
```python
import torch.nn as nn
import torch
en_layer = nn.TransformerEncoderLayer(d_model=300, nhead=2)
transformer_en = nn.TransformerEncoder(en_layer, num_layers=1)

inputs=torch.rand((10,5,300))
masks =torch.ones((5,10))
masks =masks.to(dtype=torch.bool) # same output while using dtype=torch.uint8

output=transformer_en(inputs,src_key_padding_mask=masks)
```
The above codes return output with all NAN values
![image](https://user-images.githubusercontent.com/27655180/82790940-7cdbf500-9e9f-11ea-9420-46825de90a3c.png)

**Comment by astariul** [comment]
Note for self : 

In `Transformer` class, non-zero values in `mask` are the values that will not be attended.

In my code, the mask I was using had non-zero values for positions to attend. So I simply had to inverse my mask to make it work :
```python
mask = ~mask
```

[Reference in the documentation](https://pytorch.org/docs/stable/generated/torch.nn.Transformer.html#torch.nn.Transformer.forward) :

>Note: [src/tgt/memory]_mask ensures that position i is allowed to attend the unmasked positions. If a ByteTensor is provided, the non-zero positions are not allowed to attend while the zero positions will be unchanged. If a BoolTensor is provided, positions with True are not allowed to attend while False values will be unchanged. If a FloatTensor is provided, it will be added to the attention weight. [src/tgt/memory]_key_padding_mask provides specified elements in the key to be ignored by the attention. If a ByteTensor is provided, the non-zero positions will be ignored while the zero positions will be unchanged. If a BoolTensor is provided, the positions with the value of True will be ignored while the position with the value of False will be unchanged.

**Comment by ChocoL0rd** [comment]
could someone explain my similar problem https://github.com/pytorch/pytorch/issues/119371 and what to do?
